### DISCUSSION TITLE
Slow initial load times with Next.js SSR

### DISCUSSION THREAD

**Discussion by andrew-braun** [algolia/instantsearch#5361] [Q&A]
I've got a Next.js app hosted on Vercel, mostly using static site generation (SSG), but we went with SSR for the one page that's currently using Algolia to display an initial product list and search the list. We'd initially been using CSR, but speed tests were coming upt pretty abysmal and CSR isn't the most SEO-friendly. Switching to SSR (using react-instantsearch-hooks) bumped up the speed quite a bit, but we're still seeing 5+ second initial load times, which obviously isn't ideal. 

The implementation is pretty much in line with [the Next.js section of the Algolia SSR docs. ](https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering/react-hooks/#with-nextjs)

The site is preconnecting, I've pared down the Algolia response to only the properties we're using on the frontend, components are being imported dynamically, the cache control headers are set up--and we're still getting long load times. Is there something obvious I'm missing? I'm fairly new to Algolia, so very possible there's a trick I'm not savvy to. 

I'm thinking of trying SWR or getting creative, like using ISR to regularly update a static non-Algolia page of data and then replacing it with the Algolia components when they're ready, but obviously would prefer to stick with the SSR solution Algolia provides out of the box.

**Comment by Haroenv** [comment]
Where is your Algolia server vs the server that's hosting the search? Could it be that they're physically far apart?

**Reply by andrew-braun** [reply]
Fantastic question! I'll check into that and update when I find out. Thanks :)

**Reply by andrew-braun** [reply]
Looks like Vercel's serverless functions were US East and Algolia was Singapore, so it would definitely have made sense for a second or two of latency to come from there--but unfortunately after switching Vercel to Singapore and redeploying not a ton changed on WebPageTest or PageSpeed. Mobile scores are still pretty rough, with like 6 seconds of time to interactive/largest contentful paint. Desktop scores seem to have improved a bit though (might be worth mentioning that they were actually always pretty snappy--mobile is the main issue), and definitely not a bad thing to minimize that round trip just as a general rule.

Will run some more experiments/tests with devices, locations, and various rendering strategies and try to figure out where the sweet spot is.

**Reply by Haroenv** [reply]
If you have any place we can reproduce this, that would help, six seconds for a backend render definitely is more than we regularly get.

**Reply by andrew-braun** [reply]
Sure! The repo for the site is private, but tomorrow I'll see if I can throw up a public repo with the relevant code.

In the meantime, this is the page: https://where-next-frontend-preview.vercel.app/experiences
And this is a performance test I just ran: https://www.webpagetest.org/result/220718_AiDc3Y_CEW/

Honestly, as a human, it's feeling reasonably fast to me--especially after first load--but speed-testing tools really don't like it. Though actually today we somehow managed to score 2.5 seconds _and_ 9 seconds for LCP running separate tests on the same page, so it's starting to feel more like more of an environment issue than a code issue, since these components presumably run the same way every time. 

Oh, and I feel like I have to mention that I'm maybe the 4th developer to work on this site--I'm very aware of how messy things have gotten under the hood :D

**Reply by francoischalifour** [reply]
Looking forward to getting some sandbox reproduction, because seeing the network panel, the implementation doesn't look optimal. Even though you mentioned the app was SSRed, there are 3 network requests sent to Algolia on the initial load.

**Reply by andrew-braun** [reply]
Well, this is embarassing! In the process of extracting all the necessary components for a sandbox environment, I found a gap in the prop drilling. ```serverState``` (from react-instantsearch-hooks' ```getServerState```) was coming from the parent page and being passed down to a container component, then from there to a component containing the search functionality. One of the components had a typo in the serverState it was passing, though, so it didn't end up in ```<InstantSearchSSRProvider>```. Looks like without the state it fell back to CSR rather than breaking, thus effectively negating the benefits of SSR for the page. Totally my bad there!

After fixing the gap, the page load as a human tester is _waaaaay_ better! Even on throttled connections it's coming through pretty reasonably. Speed tests are still sort of all over the place, but I inherited (and have occasionally created, case in point) a bit of a mess with this site so I think I need to do a general clean-up before doing more Algolia optimizations.

Sorry, really feel like I should have caught that typo earlier for as many times as I've been over these files! Thanks for taking the time to help--it's my first time on the Algolia/InstantSearch stack and I've been loving it so far.

**Reply by Haroenv** [reply]
Glad you have a much more performant site now, it makes me think we could have a warning if the server side rendering provider is used, but no results were passed

**Reply by andrew-braun** [reply]
Could certainly end up helping people lacking in typo-catching skills and potentially other situations as well!
